stablehlo/dialect/StablehloOps.cpp

```diff
@@ -5578,66 +5578,45 @@ LogicalResult ScatterOp::verify() {
-    if (!operandType || operandType.getRank() != 0 ||
-        !operandType.getElementType().isInteger(1))
-      return condReturnOp.emitOpError()
-             << "expects a zero-ranked tensor of i1, got "
-             << condReturnOp->getOperand(0).getType();
-  }
-  // Check the return type for the body block.
-  {
-    auto bodyReturnOp = cast<ReturnOp>(body().front().back());
-    if (bodyReturnOp->getNumOperands() != getNumOperands())
-      return bodyReturnOp.emitOpError()
-             << "expects body to return a many value as the operands ("
-             << getNumOperands() << "), got " << bodyReturnOp->getNumOperands();
-    for (const auto& enumeratedOperandTypes : llvm::enumerate(
-             llvm::zip(bodyReturnOp->getOperandTypes(), getOperandTypes()))) {
-      Type operandType = std::get<0>(enumeratedOperandTypes.value());
-      Type returnType = std::get<1>(enumeratedOperandTypes.value());
-      if (operandType != returnType)
-        return bodyReturnOp.emitOpError()
-               << "type mismatch between operand #"
-               << enumeratedOperandTypes.index()
-               << " and the enclosing WhileOp returned value: " << operandType
-               << " vs " << returnType;
-    }
-  }
+LogicalResult WhileOp::inferReturnTypeComponents(
+    MLIRContext*, Optional<Location> location, ValueShapeRange operands,
+    DictionaryAttr attributes, RegionRange regions,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  WhileOp::Adaptor adaptor(operands, attributes, regions);
+  auto operandTypes = adaptor.operand().getTypes();
+  auto condArgsTypes = adaptor.cond().front().getArgumentTypes();
+  auto bodyArgsTypes = adaptor.body().front().getArgumentTypes();
+  auto bodyReturnTypes =
+      cast<ReturnOp>(adaptor.body().front().getTerminator())->getOperandTypes();
+  if (!(hlo::isCompatibleForHloTypeInference(operandTypes, condArgsTypes)))
+    return emitOptionalError(
+        location, "expect operand matches condition block argument but got ",
+        operandTypes, " vs ", condArgsTypes);
+  if (!(hlo::isCompatibleForHloTypeInference(operandTypes, bodyArgsTypes)))
+    return emitOptionalError(
+        location, "expect operand matches body block argument but got ",
+        operandTypes, " vs ", bodyArgsTypes);
+  if (!(hlo::isCompatibleForHloTypeInference(operandTypes, bodyReturnTypes)))
+    return emitOptionalError(
+        location, "expect operand matches body block return type but got ",
+        operandTypes, " vs ", bodyReturnTypes);
+
+  auto condReturnTypes =
+      cast<ReturnOp>(adaptor.cond().front().back()).getOperandTypes();
+  if (condReturnTypes.size() != 1)
+    return emitOptionalError(
+        location, "expect condition body returns a single value but got ",
+        condReturnTypes.size());
+  auto operandType = condReturnTypes[0].dyn_cast<RankedTensorType>();
+  if (!operandType || operandType.getRank() != 0 ||
```

Comment:
Why reject unranked returns? If the condition returns `tensor<*xi1>`, I think it's just fine.

Reply:
Here is how I think of it: From the XLA implementation: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/shape_inference.cc#L2882 it compare it with `ShapeUtil::MakeShape(PRED, {})`, and regular dims are not ignored https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/shape_util.cc#L725

From my understanding, the expected return value of cond is a 0-ranked i1 tensor ([PRED in XLA), but *xi1 is unclear here. If *x is not scalar and have several l 1s and 0s, then will the while loop exit or not? 

Besides, this is same code as MHLO. So this strict check will not fail any MHLO tests.

Reply:
Yes this sounds reasonable to me. Add another commit to tolerant tensor<*xi1>. Also twist the unranked test.